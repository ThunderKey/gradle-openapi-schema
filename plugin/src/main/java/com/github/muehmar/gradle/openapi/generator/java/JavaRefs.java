package com.github.muehmar.gradle.openapi.generator.java;

public class JavaRefs {
  private JavaRefs() {}

  public static final String LOCAL_DATE = "java.time.LocalDate";
  public static final String OPTIONAL = "java.util.Optional";
}
