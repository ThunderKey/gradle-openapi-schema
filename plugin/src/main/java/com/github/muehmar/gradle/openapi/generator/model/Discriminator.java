package com.github.muehmar.gradle.openapi.generator.model;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class Discriminator {
  private final Name propertyName;
  private final Optional<Map<String, PojoName>> mapping;

  private Discriminator(Name propertyName, Optional<Map<String, PojoName>> mapping) {
    this.propertyName = propertyName;
    this.mapping = mapping;
  }

  public static Discriminator fromPropertyName(Name propertyName) {
    return new Discriminator(propertyName, Optional.empty());
  }

  public static Discriminator fromPropertyNameAndMapping(
      Name propertyName, Map<String, PojoName> mapping) {
    return new Discriminator(propertyName, Optional.of(mapping));
  }

  public Discriminator withMapping(Optional<Map<String, PojoName>> mapping) {
    return new Discriminator(propertyName, mapping);
  }

  public Name getPropertyName() {
    return propertyName;
  }

  public String getValueForPojoName(PojoName pojoName) {
    return mapping.orElse(Collections.emptyMap()).entrySet().stream()
        .filter(e -> e.getValue().equals(pojoName))
        .findFirst()
        .map(Map.Entry::getKey)
        .orElse(pojoName.getName().asString());
  }
}
