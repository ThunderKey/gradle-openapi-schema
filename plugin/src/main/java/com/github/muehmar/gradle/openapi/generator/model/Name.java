package com.github.muehmar.gradle.openapi.generator.model;

import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;

/** Name of a pojo or a pojo member. */
@EqualsAndHashCode
public class Name {
  private final String value;

  private Name(String val) {
    this.value = val;
  }

  public static Name ofString(String val) {
    if (val == null || val.trim().isEmpty()) {
      throw new IllegalArgumentException("A name must not be null or empty");
    }

    return new Name(val);
  }

  public String asString() {
    return value;
  }

  public Name map(UnaryOperator<String> f) {
    return Name.ofString(f.apply(value));
  }

  public Name append(String append) {
    return new Name(value + append);
  }

  public Name append(Name append) {
    return new Name(value + append.value);
  }

  public Name prefix(String prefix) {
    return new Name(prefix + value);
  }

  public SuffixedName suffix(String suffix) {
    return SuffixedName.ofNameAndSuffix(this, suffix);
  }

  public boolean equalsIgnoreCase(Name other) {
    return value.equalsIgnoreCase(other.value);
  }

  public Name startUpperCase() {
    return map(s -> s.substring(0, 1).toUpperCase() + s.substring(1));
  }

  @Override
  public String toString() {
    return asString();
  }
}
