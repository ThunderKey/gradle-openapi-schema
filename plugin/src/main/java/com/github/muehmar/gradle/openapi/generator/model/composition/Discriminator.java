package com.github.muehmar.gradle.openapi.generator.model.composition;

import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class Discriminator {
  private final Name propertyName;
  private final DiscriminatorType type;
  private final Optional<Map<String, Name>> mapping;

  private Discriminator(
      Name propertyName, DiscriminatorType type, Optional<Map<String, Name>> mapping) {
    this.propertyName = propertyName;
    this.type = type;
    this.mapping = mapping;
  }

  public static Discriminator typeDiscriminator(
      UntypedDiscriminator untypedDiscriminator, DiscriminatorType type) {
    return new Discriminator(
        untypedDiscriminator.getPropertyName(), type, untypedDiscriminator.getMapping());
  }

  public Name getPropertyName() {
    return propertyName;
  }

  public DiscriminatorType getType() {
    return type;
  }

  public String getValueForSchemaName(Name schemaName) {
    return mapping.orElse(Collections.emptyMap()).entrySet().stream()
        .filter(e -> e.getValue().equals(schemaName))
        .findFirst()
        .map(Map.Entry::getKey)
        .orElse(schemaName.asString());
  }
}
