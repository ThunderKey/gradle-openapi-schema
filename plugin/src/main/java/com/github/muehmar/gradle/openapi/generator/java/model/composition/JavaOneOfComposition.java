package com.github.muehmar.gradle.openapi.generator.java.model.composition;

import static com.github.muehmar.gradle.openapi.generator.java.model.composition.Assertion.assertAllObjectPojos;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojo;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.composition.OneOfComposition;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaOneOfComposition {
  private final NonEmptyList<JavaObjectPojo> pojos;
  private final Optional<Discriminator> discriminator;

  private JavaOneOfComposition(
      NonEmptyList<JavaPojo> pojos, Optional<Discriminator> discriminator) {
    this.pojos = assertAllObjectPojos(pojos);
    this.discriminator = discriminator;
  }

  public static JavaOneOfComposition wrap(
      OneOfComposition oneOfComposition, PojoType type, TypeMappings typeMappings) {
    return new JavaOneOfComposition(
        oneOfComposition
            .getPojos()
            .map(pojo -> JavaPojo.wrap(pojo, typeMappings))
            .map(result -> result.getTypeOrDefault(type)),
        oneOfComposition.getDiscriminator());
  }

  public static JavaOneOfComposition fromPojos(NonEmptyList<JavaPojo> pojos) {
    return new JavaOneOfComposition(pojos, Optional.empty());
  }

  public static JavaOneOfComposition fromPojosAndDiscriminator(
      NonEmptyList<JavaPojo> pojos, Discriminator discriminator) {
    return new JavaOneOfComposition(pojos, Optional.of(discriminator));
  }

  public NonEmptyList<JavaObjectPojo> getPojos() {
    return pojos;
  }

  public PList<JavaPojoMember> getMembers() {
    return pojos
        .toPList()
        .flatMap(JavaObjectPojo::getAllMembersForComposition)
        .map(JavaPojoMember::asOneOfMember);
  }

  public Optional<Discriminator> getDiscriminator() {
    return discriminator;
  }
}
