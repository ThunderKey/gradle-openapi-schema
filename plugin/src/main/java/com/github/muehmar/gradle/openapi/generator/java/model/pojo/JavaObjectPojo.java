package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ConstructorContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaIdentifier;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.wrap(type.mapName(objectPojo.getName())),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(comp -> JavaOneOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(name.getSchemaName());
  }

  @Override
  public JavaIdentifier getClassName() {
    return name.asJavaName().asIdentifier();
  }

  public JavaIdentifier prefixedClassNameForMethod(String prefix) {
    if (prefix.isEmpty()) {
      return name.asJavaName().startLowerCase().asIdentifier();
    } else {
      return name.asJavaName().startUpperCase().prefix(prefix).asIdentifier();
    }
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  private PList<JavaPojoMember> getAllOfMembers() {
    return allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getOneOfMembers() {
    return oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
  }

  private PList<JavaPojoMember> getAnyOfMembers() {
    return anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    return getAllOfMembers()
        .concat(getOneOfMembers())
        .concat(getAnyOfMembers())
        .distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return MemberContentBuilder.create()
        .isArrayPojo(false)
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return HashCodeContentBuilder.create()
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return EqualsContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return ToStringContentBuilder.create()
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return ConstructorContentBuilder.create()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return WitherContentBuilder.create()
        .className(getClassName())
        .membersForWithers(getMembers())
        .membersForConstructorCall(getAllMembers())
        .andAllOptionals()
        .additionalProperties(additionalProperties)
        .build();
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
