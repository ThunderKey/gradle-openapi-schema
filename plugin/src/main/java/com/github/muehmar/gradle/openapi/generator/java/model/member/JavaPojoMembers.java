package com.github.muehmar.gradle.openapi.generator.java.model.member;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import java.util.function.UnaryOperator;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * The members of a {@link JavaObjectPojo}. This container ensures, that the same property does not
 * exist twice. Adding a property which is already present will not add a new property/member, it
 * will keep the current property but maybe with lower restrictions concerning {@link Nullability}
 * and {@link Necessity}.
 */
@EqualsAndHashCode
@ToString
public class JavaPojoMembers {
  private final PList<JavaPojoMember> members;

  private JavaPojoMembers(PList<JavaPojoMember> members) {
    this.members = members;
  }

  public static JavaPojoMembers empty() {
    return new JavaPojoMembers(PList.empty());
  }

  public static JavaPojoMembers fromMembers(PList<JavaPojoMember> members) {
    return JavaPojoMembers.empty().add(members);
  }

  public JavaPojoMembers add(JavaPojoMember memberToAdd) {
    final boolean hasMemberWithSameKey =
        members.exists(member -> member.getMemberKey().equals(memberToAdd.getMemberKey()));
    if (hasMemberWithSameKey) {
      final PList<JavaPojoMember> newMembers =
          members.map(member -> member.mergeToLeastRestrictive(memberToAdd).orElse(member));
      return new JavaPojoMembers(newMembers);
    } else {
      return new JavaPojoMembers(members.add(memberToAdd));
    }
  }

  public JavaPojoMembers add(PList<JavaPojoMember> membersToAdd) {
    return membersToAdd.foldLeft(this, JavaPojoMembers::add);
  }

  public JavaPojoMembers add(JavaPojoMembers membersToAdd) {
    return add(membersToAdd.asList());
  }

  public JavaPojoMembers map(UnaryOperator<JavaPojoMember> f) {
    return new JavaPojoMembers(members.map(f));
  }

  public PList<JavaPojoMember> asList() {
    return members;
  }

  public boolean isEmpty() {
    return members.isEmpty();
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size();
  }
}
