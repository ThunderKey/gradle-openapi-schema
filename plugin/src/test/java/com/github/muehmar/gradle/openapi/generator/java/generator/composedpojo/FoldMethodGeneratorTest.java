package com.github.muehmar.gradle.openapi.generator.java.generator.composedpojo;

import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo1;
import static com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos.sampleObjectPojo2;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import ch.bluecare.commons.data.NonEmptyList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.model.Discriminator;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import java.util.HashMap;
import java.util.function.Function;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(SnapshotExtension.class)
class FoldMethodGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("OneOfNoDiscriminator")
  void generate_when_calledWithoutDiscriminator_then_correctContent() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        FoldMethodGenerator.foldMethodGenerator();
    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(sampleObjectPojo1(), sampleObjectPojo2()),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_FUNCTION::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_SUPPLIER::equals));
    assertEquals(2, writer.getRefs().distinct(Function.identity()).size());
  }

  @Test
  @SnapshotName("OneOfDiscriminatorWithoutMapping")
  void generate_when_calledWithDiscriminator_then_correctContent() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        FoldMethodGenerator.foldMethodGenerator();
    final Discriminator discriminator =
        Discriminator.fromPropertyName(JavaPojoMembers.requiredString().getName().asName());
    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfComposition.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1(), sampleObjectPojo2()), discriminator);

    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(javaOneOfComposition),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_FUNCTION::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_SUPPLIER::equals));
    assertEquals(2, writer.getRefs().distinct(Function.identity()).size());
  }

  @Test
  @SnapshotName("OneOfDiscriminatorWithMapping")
  void generate_when_calledWithDiscriminatorAndMapping_then_correctContent() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        FoldMethodGenerator.foldMethodGenerator();
    final JavaObjectPojo sampleObjectPojo1 = sampleObjectPojo1();
    final JavaObjectPojo sampleObjectPojo2 = sampleObjectPojo2();

    final HashMap<String, Name> mapping = new HashMap<>();
    mapping.put("obj1", sampleObjectPojo1.getSchemaName().asName());
    mapping.put("obj2", sampleObjectPojo2.getSchemaName().asName());
    final Discriminator discriminator =
        Discriminator.fromPropertyNameAndMapping(
            JavaPojoMembers.requiredString().getName().asName(), mapping);

    final JavaOneOfComposition javaOneOfComposition =
        JavaOneOfComposition.fromPojosAndDiscriminator(
            NonEmptyList.of(sampleObjectPojo1, sampleObjectPojo2), discriminator);

    final Writer writer =
        generator.generate(
            JavaPojos.oneOfPojo(javaOneOfComposition),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_FUNCTION::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_SUPPLIER::equals));
    assertEquals(2, writer.getRefs().distinct(Function.identity()).size());
  }

  @Test
  @SnapshotName("AnyOf")
  void generate_when_anyOf_then_correctOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        FoldMethodGenerator.foldMethodGenerator();

    final Writer writer =
        generator.generate(
            JavaPojos.anyOfPojo(sampleObjectPojo1(), sampleObjectPojo2()),
            TestPojoSettings.defaultSettings(),
            Writer.createDefault());

    expect.toMatchSnapshot(writer.asString());
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_LIST::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_ARRAY_LIST::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_FUNCTION::equals));
    assertEquals(3, writer.getRefs().distinct(Function.identity()).size());
  }
}
