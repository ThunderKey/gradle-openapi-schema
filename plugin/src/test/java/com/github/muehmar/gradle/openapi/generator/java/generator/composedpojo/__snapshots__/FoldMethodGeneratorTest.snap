AnyOf=[

public <T> List<T> fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto) {
  final List<T> result = new ArrayList<>();
  if (isValidAgainstUser()) {
    result.add(onUserDto.apply(asUserDto()));
  }
  if (isValidAgainstNecessityAndNullability()) {
    result.add(onNecessityAndNullabilityDto.apply(asNecessityAndNullabilityDto()));
  }
  return result;
}
]


OneOfDiscriminatorWithMapping=[
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto, Supplier<T> onInvalid) {
  if ("UserValue".equals(requiredStringVal) && isValidAgainstUser()) {
    return onUserDto.apply(asUserDto());
  }
  else if ("NNVariantsValue".equals(requiredStringVal) && isValidAgainstNecessityAndNullability()) {
    return onNecessityAndNullabilityDto.apply(asNecessityAndNullabilityDto());
  }
  else {
    return onInvalid.get();
  }
}

/**
 * Folds this instance using the given mapping functions for the DTO's. If this
 * instance is valid against exactly one of the specified schemas, its
 * corresponding mapping function gets executed with the DTO as input and its
 * result is returned.<br><br>
 * 
 * I.e. if the JSON was valid against the schema 'User', the mapping method {@code
 * onUserDto} gets executed with the {@link UserDto} as argument.<br><br>
 * 
 * This method assumes this instance is either manually or automatically
 * validated, i.e. the JSON is valid against exactly one of the schemas. If it is
 * either valid against no schema or multiple schemas, it will throw an {@link
 * IllegalStateException}.
 */
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto) {
  return fold(
    onUserDto,
    onNecessityAndNullabilityDto,
    () -> {throw new IllegalStateException("Unable to fold ComposedOneofDto: Not valid against one of the schemas [UserDto, NecessityAndNullabilityDto].");}
  );
}
]


OneOfDiscriminatorWithoutMapping=[
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto, Supplier<T> onInvalid) {
  if ("User".equals(requiredStringVal) && isValidAgainstUser()) {
    return onUserDto.apply(asUserDto());
  }
  else if ("NecessityAndNullability".equals(requiredStringVal) && isValidAgainstNecessityAndNullability()) {
    return onNecessityAndNullabilityDto.apply(asNecessityAndNullabilityDto());
  }
  else {
    return onInvalid.get();
  }
}

/**
 * Folds this instance using the given mapping functions for the DTO's. If this
 * instance is valid against exactly one of the specified schemas, its
 * corresponding mapping function gets executed with the DTO as input and its
 * result is returned.<br><br>
 * 
 * I.e. if the JSON was valid against the schema 'User', the mapping method {@code
 * onUserDto} gets executed with the {@link UserDto} as argument.<br><br>
 * 
 * This method assumes this instance is either manually or automatically
 * validated, i.e. the JSON is valid against exactly one of the schemas. If it is
 * either valid against no schema or multiple schemas, it will throw an {@link
 * IllegalStateException}.
 */
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto) {
  return fold(
    onUserDto,
    onNecessityAndNullabilityDto,
    () -> {throw new IllegalStateException("Unable to fold ComposedOneofDto: Not valid against one of the schemas [UserDto, NecessityAndNullabilityDto].");}
  );
}
]


OneOfNoDiscriminator=[
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto, Supplier<T> onInvalid) {
  if (isValidAgainstUser()) {
    return onUserDto.apply(asUserDto());
  }
  else if (isValidAgainstNecessityAndNullability()) {
    return onNecessityAndNullabilityDto.apply(asNecessityAndNullabilityDto());
  }
  else {
    return onInvalid.get();
  }
}

/**
 * Folds this instance using the given mapping functions for the DTO's. If this
 * instance is valid against exactly one of the specified schemas, its
 * corresponding mapping function gets executed with the DTO as input and its
 * result is returned.<br><br>
 * 
 * I.e. if the JSON was valid against the schema 'User', the mapping method {@code
 * onUserDto} gets executed with the {@link UserDto} as argument.<br><br>
 * 
 * This method assumes this instance is either manually or automatically
 * validated, i.e. the JSON is valid against exactly one of the schemas. If it is
 * either valid against no schema or multiple schemas, it will throw an {@link
 * IllegalStateException}.
 */
public <T> T fold(Function<UserDto, T> onUserDto, Function<NecessityAndNullabilityDto, T> onNecessityAndNullabilityDto) {
  return fold(
    onUserDto,
    onNecessityAndNullabilityDto,
    () -> {throw new IllegalStateException("Unable to fold ComposedOneofDto: Not valid against one of the schemas [UserDto, NecessityAndNullabilityDto].");}
  );
}
]